api-documentation/shacl.md

```diff
@@ -0,0 +1,181 @@
+
+```json
+{
+  "@id": "mov:NewMovieShape",
+  "@type": "sh:NodeShape",
+  "sh:targetClass": "mov:Movie",
+  "sh:property": [{
+    "sh:path": "schema:name",
+    "sh:or": [{
+      "sh:datatype": "xsd:string"
+    }, {
+      "sh:datatype": "rdf:langString"
+    }],
+    "sh:name": "Title",
+    "sh:description": "Movie's title",
+    "sh:minCount": 1
+  }, {
+    "sh:path": "mov:director",
+    "sh:class": "schema:Person",
+    "sh:nodeKind": "sh:IRI",
+    "sh:maxCount": 1
+  }, {
+    "sh:path": "schema:genre",
+    "sh:in": [
+      "Comedy",
+      "Drama",
+      "Documentary"
+    ]
+  }]
+}
+```
+
+While the Shape repeats some of the information typically found on the supported class `mov:Movie` it is clear that it can also provide much more information about the expected payload of the request to create a movie resource, which would be otherwise impossible with just Hydra Core terms:
+
+1. About `schema:name` property:
+   * At least one value is required
+   * There can be multiple values
+   * Values can be a plain string of string with language tag
+2. About `mov:director` property:
+   * Must be a resource identifier of a `schema:Person` resource
+   * There can be at most one value
+3. About `schema:genre` property:
+   * Values should be one of those provided by the `sh:in` enumeration
+
+{% hint style='info' %}
+The only gray area is the director property, where it is up to the client to figure out the available instances of `schema:Person`
+{% endhint %}
+
+### Describing IRI templates
+
+Another way for a client to create HTTP requests is by filling an IRI template with string values and dereferencing the resulting identifier. This works well for example for filtering a collection with query strings.
+
+An instance of tht `mov:Movies` class above could provide such a search template to filter movies by genre.
+
+```json
+{
+  "@id": "movies",
+  "@type": ["Collection", "mov:Movies"],
+  "search": {
+    "@type": "IriTemplate",
+    "template": "movies{?genre}",
+    "mapping": [{
+      "@type": "IriTemplateMapping",
+      "variable": "slug",
+      "property": "schema:genre"
+    }]
+  }
+}
+```
+
+While this is enough to match an existing resource which has the `schema:genre` property, it is too little information to gather user input, such as with a form on a web page.
+
+To enrich the template definition, a SHACL shape can be added to further specify the template mappings.
+
+```diff
+{
+   "@id": "movies",
+   "@type": ["Collection", "mov:Movies"],
+   "search": {
+     "@type": "IriTemplate",
+     "template": "movies{?genre}",
+     "mapping": [{
+       "@type": "IriTemplateMapping",
+       "variable": "slug",
+       "property": "schema:genre"
+-    }]
++    }],
++    "hashi:shape": {
++      "@type": "sh:Shape",
++      "rdfs:label": "Search movies collection",
++      "sh:property": [{
++        "sh:path": "schema:genre",
++        "sh:name": "Genre",
```

Comment:
I believe sh:name can be replaced with either `hydra:title` or just `rdfs:label'

Reply:
`hydra:title` has no meaning in SHACL context

`sh:name` is SHACL's method for annotating `PropertyShape`

I know it's weird but decided that way because `rdfs:label` would effectively annotate the Shape where the intention is to indirectly annotate the property :shrug:

Reply:
Ok, but which label client should use i.e. to build an UI?

Reply:
Of what choice? There is only one label here (`sh:name`) :)

Reply:
I would consider a Shape as separate concerns.

So, the API describes its hypermedia with a Shape. That Shape can be then passed to a component which understand SHACL to build the UI for example and run validation. That SHACL-specific component would not be aware of Hydra though, hence I stick to only terms defined in SHACL.

The Shape is its own independent description which a Hydra client can take advantage of